File: 69. Sqrt(x)/Solution.py
class Solution:
    # @param x, an integer
    # @return an integer
    def mySqrt(self, x):
        if x < 0:
            return -1
        if x == 0:
            return 0
            
        left = 0
        right = x / 2 + 1
        while left <= right:
            mid = (left + right) / 2
            if mid * mid == x:
                return mid
            elif mid * mid > x:
                right = mid - 1
            else :
                left = mid + 1
        return right
        

File: 69. Sqrt(x)/Solution.java
// LeetCode #69: Sqrt(x)

/*
Implement int sqrt(int x).

Compute and return the square root of x.
*/

public class Solution {
    public int mySqrt(int x) {
        if (x < 0)
            return -1;
        if (x == 0)
            return 0;
            
        long left = 0;
        long right = x / 2 + 1; // sqrt(x) < x/2 + 1
        
        while (left <= right) {
            long mid = (left + right) / 2;
            long square = mid * mid;
            if (square == x)
                return (int)mid;
            else if (square > x)
                right = mid - 1;
            else
                left = mid + 1;
        }
        return (int)right;
    }
}